flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/checkpointing/BarrierAlignmentUtilTest.java

```
@@ -0,0 +1,69 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io.checkpointing;
+
+import org.apache.flink.api.common.operators.MailboxExecutor;
+import org.apache.flink.streaming.runtime.tasks.StreamTaskActionExecutor;
+import org.apache.flink.streaming.runtime.tasks.TestProcessingTimeService;
+import org.apache.flink.streaming.runtime.tasks.mailbox.MailboxExecutorImpl;
+import org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor;
+import org.apache.flink.streaming.runtime.tasks.mailbox.TaskMailbox;
+import org.apache.flink.streaming.runtime.tasks.mailbox.TaskMailboxImpl;
+
+import org.junit.Test;
+
+import java.io.IOException;
+import java.time.Duration;
+
+import static org.junit.Assert.assertThrows;
+
+public class BarrierAlignmentUtilTest {
+
+    @Test
+    public void testDelayableTimerNotHiddenException() throws Exception {
+        TaskMailbox mailbox = new TaskMailboxImpl();
+        MailboxProcessor mailboxProcessor =
+                new MailboxProcessor(controller -> {}, mailbox, StreamTaskActionExecutor.IMMEDIATE);
+        MailboxExecutor mailboxExecutor =
+                new MailboxExecutorImpl(
+                        mailbox, 0, StreamTaskActionExecutor.IMMEDIATE, mailboxProcessor);
+
+        TestProcessingTimeService timerService = new TestProcessingTimeService();
+        timerService.setCurrentTime(System.currentTimeMillis());
+
+        BarrierAlignmentUtil.DelayableTimer delayableTimer =
+                BarrierAlignmentUtil.createRegisterTimerCallback(mailboxExecutor, timerService);
+
+        Duration delay = Duration.ofMinutes(10);
+
+        delayableTimer.registerTask(
+                () -> {
+                    // simulate IOException in checkpoint sync phase
+                    throw new IOException();
```

nit: I would throw a custom exception, like for example `org.apache.flink.runtime.operators.testutils.ExpectedTestException`. It's less likely to collide with some actual failure reported via `IOException`.